persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/DiscreteCopyNumberSparkRepository.java

```diff
@@ -0,0 +1,207 @@
+        sb.append("FROM cna ");
+        sb.append("GROUP BY " + mainCol + ", VALUE ");
+        sb.append("SORT BY " + mainCol);
+
+        Dataset<Row> result = spark.sql(sb.toString());
+
+        List<CopyNumberCountByGene> copyNumberCountByGenes = result.collectAsList().stream()
+            .map(r -> mapToCopyNumberCountByGene(r, mainCol)).collect(Collectors.toList());
+
+        return getGeneDetails(copyNumberCountByGenes, mainCol, entrezGeneIds);
+    }
+
+    @Override
+    public List<CopyNumberCountByGene> getPatientCountByGeneAndAlterationAndPatientIds(String molecularProfileId, List<String> patientIds, List<Integer> entrezGeneIds, List<Integer> alterations) {
+
+        Dataset<Row> casels = spark.read().parquet(PARQUET_DIR + ParquetConstants.CASE_LIST_DIR + molecularProfileId);
+        casels = casels.withColumnRenamed("case_list_ids", "SAMPLE_ID");
+
+        // get studyId
+        List<String> studyIds = casels.select("cancer_study_identifier").distinct()
+            .collectAsList().stream().map(r -> r.getString(0)).collect(Collectors.toList());
+        String studyId = studyIds.get(0);
+        // filter by alterations
+        Dataset<Row> cnaData = spark.read().parquet(PARQUET_DIR + ParquetConstants.STUDIES_DIR + studyId + "/" + ParquetConstants.DATA_CNA);
+        if (!CollectionUtils.isEmpty(alterations)) {
+            cnaData = cnaData.where(cnaData.col("VALUE").isin(alterations.toArray()));
+        }
+        // filter by patientIds
+        Dataset<Row> clinicalSample = spark.read().parquet(PARQUET_DIR + ParquetConstants.STUDIES_DIR + studyId + "/" + ParquetConstants.CLINICAL_SAMPLE);
+        if (!CollectionUtils.isEmpty(patientIds)) {
+            clinicalSample = clinicalSample
+                .where(clinicalSample.col("PATIENT_ID").isin(patientIds.toArray()));
+        }
+        casels = casels.join(clinicalSample, "SAMPLE_ID");
+        cnaData = cnaData.join(casels, "SAMPLE_ID");
+
+        String mainCol = Arrays.asList(cnaData.columns()).contains("Entrez_Gene_Id") ? "Entrez_Gene_Id" : "Hugo_Symbol";
+        cnaData.createOrReplaceTempView("cna");
+        StringBuilder sb = new StringBuilder();
+        sb.append("SELECT " + mainCol + ", VALUE, COUNT(DISTINCT(PATIENT_ID)) as numberOfAlteredCases ");
+        sb.append("FROM cna ");
+        sb.append("GROUP BY " + mainCol + ", VALUE ");
+        sb.append("SORT BY " + mainCol);
+
+        Dataset<Row> result = spark.sql(sb.toString());
+
+        List<CopyNumberCountByGene> copyNumberCountByGenes = result.collectAsList().stream()
+            .map(r -> mapToCopyNumberCountByGene(r, mainCol)).collect(Collectors.toList());
+
+        return getGeneDetails(copyNumberCountByGenes, mainCol, entrezGeneIds);
+    }
+    
+    private List<CopyNumberCountByGene> getGeneDetails(List<CopyNumberCountByGene> copyNumberCountByGenes, String colName, List<Integer> entrezGeneIds) {
+        if ("Hugo_Symbol".equalsIgnoreCase(colName)) {
```

Comment:
both if and else blocks looks similar, refactoring should be done.

Reply:
logic for fetching gene detail is now in ParquetWriter per your suggestion.